### PR TITLE
HPCC-11962 Monitor Dali and kill (after timeout) process

### PR DESCRIPTION
**Pull request by jakesmith**
Forcibly kill eclagent process if Dali has closed and process
doesn't exit gracefully.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested by:
1) running system
2) stopping thor,
3) running Thor job (so blocked on Thor).
4) stopping system.

That left the eclagent process orphaned.

With this PR, it spotted Dali had gone down, and after the pre-requisite 60 second timeout,
it killed itself.

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

File: ecl/eclagent/eclagent.cpp
Comment (mckellyln) [inline ecl/eclagent/eclagent.cpp]: just being picky, I prefer raise() if sending to oneself.

Comment (jakesmith) [reply ecl/eclagent/eclagent.cpp]: I don't think we use raise() anywhere at the moment.
Above is copy and paste of code that we use in quite a few place.

I thought about adding a wrapper to remove the platform #ifdef's here, but didn't.
We can do that another time and change to raise()  (so everywhere uses).

**Comment by mckellyln** [comment]
@jakesmith looks good.  Have you tested it ?

**Comment by jakesmith** [comment]
>Have you tested it ?

Yes, have updated the Testing section above.

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
@jakesmith looks good to merge.

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

File: ecl/eclagent/eclagent.cpp
Comment (richardkchapman) [reply ecl/eclagent/eclagent.cpp]: Seems like this is likely to block the very thread that is trying to close cleanly?

Also, isn't there a guillotine thread running that (a) should have killed eventually anyway and (b) could check periodically that dali connection is still valid.

**Comment by richardkchapman** [comment]
@jakesmith Some questions

**Comment by jakesmith** [comment]
>Also, isn't there a guillotine thread running that (a) should have killed eventually anyway and (b) could check periodically that dali connection is still valid.

There is a guillotine mechanism, set via workunit option "maxRunTime", but it's not enabled by default, probably because eclagent processes can be blocked by various things for undefined periods.

**Comment by jakesmith** [comment]
>Seems like this is likely to block the very thread that is trying to close cleanly?

This onClose is called by a connection monitor thread inside MP, it will be only be invoked when Dali has closed. When it does, it will do a sem.wait for up to 1 minute, if the eclagent process is already existing then the dtor will signal and this thread will do nothing, if 1 minute expires (after being told Dali went down), then the process will be forcibly killed.

**Comment by jakesmith** [comment]
@richardkchapman - added some responses to questions.

**Comment by richardkchapman** [comment]
@ghalliday Please sanity-check

**Comment by ghalliday** [comment]
@jakesmith is there a danger that this will cause deadlock/race e.g.:

- a connection monitor which is registered after this one will never get called.  (e.g. querySessionManager() called late for some reason)
- An (unlikely) race condition if the semaphore is signalled, but onClose() times out - WARNLOG will crash if the process has exited.
-

**Comment by jakesmith** [comment]
>a connection monitor which is registered after this one will never get called. (e.g. querySessionManager() called late for some reason)

I don't think that's a problem, it's likely querySessionManager() will have been called earlier, but even it wasn't, if it was use late (after signal that Dali had gone had been received), then it couldn't be used, or rather those uses would stall or error. 

>An (unlikely) race condition if the semaphore is signalled, but onClose() times out - WARNLOG will crash if the process has exited.

Very unlikely I think, it has a [relatively] long way to go before logging is defunct., it should have got past that line by that time.
But yes theoretically there's a window, I'll change so it avoids it.

**Comment by jakesmith** [comment]
@ghalliday  - Actually, I take it back, I don't think there is a window.
If the it's signaled and process is winding up (and in or about to do WARNLOG), it must be before it has stopped MP, which will have stopped (joined) the thread that calls the IConnectionMonitor::onClose methods.
So afaics, it cannot be in onClose dealing with tracing after it is invalid

**Comment by ghalliday** [comment]
@richardkchapman I'll let you take a last look and merge

**Comment by jakesmith** [comment]
@richardkchapman - agree, have switched it to master.

**Pull request merged**
